doc/source/raysgd/v2/user_guide.rst

```diff
@@ -745,21 +745,89 @@ A couple caveats:
 
     analysis = tune.run(trainable, config={
         "lr": tune.uniform(), "batch_size": tune.randint(1, 2, 3)}, num_samples=12)
-..
-    Advanced APIs
-    -------------
 
-    TODO
+.. _sgd-advanced-api:
 
-    Training Run Iterator API
-    ~~~~~~~~~~~~~~~~~~~~~~~~~
+Advanced APIs
+-------------
+In addition to the standard ``Trainer.run`` API, there are two additional ways
+to run your distributed training code:
+1. An iterator API to have more customization of the training loop
+2. A stateful class API for customization on the worker state.
 
-    TODO
+.. _sgd-iterator:
 
-    Stateful Class API
-    ~~~~~~~~~~~~~~~~~~
+Training Run Iterator API
+~~~~~~~~~~~~~~~~~~~~~~~~~
+The ``Trainer.run_iterator`` API returns a :ref:`sgd-api-iterator` object that allows you to
+manually iterate through the intermediate results from ``sgd.report()`` in your training script.
+This could be useful to have more customization of the training/validation loop or if your training logic
+cannot fit into a single training function to be executed in one-shot.
 
-    TODO
+.. code-block:: python
```

Comment:
Can you make these examples runnable? Pseudocode is great for high level understanding but from my experience a simple but concrete example is more effective when I'm actually trying to learn/use a feature.

Reply:
In particular:
1. the iterator API example should highlight something that callbacks cannot support.
2. the class API example should use state.